builder/pipeline.cc

```diff
@@ -795,55 +913,148 @@ class pipeline_builder {
   //   For each of the new loops, the `build()` is called for the case when there
   //   are func which need to be produced in that new loop.
   stmt build(const stmt& body, const func* base_f, const loop_id& at) {
-    std::vector<stmt> results;
-
+    symbol_map<var> uncropped_subs;
+    std::vector<std::tuple<stmt, int, int>> results;
     // Build the functions computed at this loop level.
     for (auto i = order_.rbegin(); i != order_.rend(); ++i) {
       const func* f = *i;
       const auto& compute_at = compute_at_levels_.find(f);
       assert(compute_at != compute_at_levels_.end());
+      std::set<var> old_candidates = candidates_for_allocation_[at];
 
       const auto& realize_at = realization_levels_.find(f);
       assert(realize_at != realization_levels_.end());
 
       if (compute_at->second == at && !f->loops().empty()) {
-        results.push_back(make_loops(f));
+        std::tuple<stmt, int, int> f_body = make_loops(f);
+        // This is a special case for the buffers which are produced and consumed inside
+        // of this loop. In this case we simply wrap loop body with corresponding allocations.
+        if (candidates_for_allocation_[at].size() > old_candidates.size() + 1) {
+          std::set<var> to_remove;
+          for (auto b : candidates_for_allocation_[at]) {
+            if (old_candidates.count(b) > 0) continue;
+            if (allocation_info_[b]->consumers_produced != allocation_info_[b]->deps_count) continue;
```

Comment:
Make a reference to the struct here. This if body does the same map lookup 7 times!